### ISSUE TITLE
Unable to connect AVRO table to Kafka-connect JDBC sink

### ISSUE THREAD

**Issue by fortinj66** [confluentinc/ksql#2250]
After getting the ksql table create issue figuredout I am now facing a new issue...

I am trying to use kafka-connect-jdbc-sink  to get the ksql table (in avro format) imported into mysql.  It would seem to be a pretty easy thing to do...

Unfortunately, I am running into issue,  I am getting the following from the connector:

```
[2018-12-08 17:30:53,384] INFO Checking MySql dialect for existence of table "DSB_ERROR_TABLE_WINDOWED" (io.confluent.connect.jdbc.dialect.MySqlDatabaseDialect)
[2018-12-08 17:30:53,385] INFO Using MySql dialect table "DSB_ERROR_TABLE_WINDOWED" present (io.confluent.connect.jdbc.dialect.MySqlDatabaseDialect)

[2018-12-08 17:30:53,388] INFO Setting metadata for table "DSB_ERROR_TABLE_WINDOWED" to Table{name='"DSB_ERROR_TABLE_WINDOWED"', columns=[Column{'MOD_CLASS', isPrimaryKey=false, allowsNull=true, sqlType=VARCHAR}, Column{'METHOD', isPrimaryKey=false, allowsNull=true, sqlType=VARCHAR}, Column{'KSQL_COL_2', isPrimaryKey=false, allowsNull=true, sqlType=BIGINT}]} (io.confluent.connect.jdbc.util.TableDefinitions)


[2018-12-08 17:30:53,390] ERROR WorkerSinkTask{id=dev-dsb-errors-mysql-sink-0} Task threw an uncaught and unrecoverable exception. Task is being killed and will not recover until manually restarted. (org.apache.kafka.connect.runtime.WorkerSinkTask)
org.apache.kafka.connect.errors.ConnectException: No fields found using key and value schemas for table: DSB_ERROR_TABLE_WINDOWED
        at io.confluent.connect.jdbc.sink.metadata.FieldsMetadata.extract(FieldsMetadata.java:127)
        at io.confluent.connect.jdbc.sink.metadata.FieldsMetadata.extract(FieldsMetadata.java:64)
        at io.confluent.connect.jdbc.sink.BufferedRecords.add(BufferedRecords.java:79)
        at io.confluent.connect.jdbc.sink.BufferedRecords.add(BufferedRecords.java:124)
        at io.confluent.connect.jdbc.sink.JdbcDbWriter.write(JdbcDbWriter.java:63)
        at io.confluent.connect.jdbc.sink.JdbcSinkTask.put(JdbcSinkTask.java:75)
        at org.apache.kafka.connect.runtime.WorkerSinkTask.deliverMessages(WorkerSinkTask.java:564)
        at org.apache.kafka.connect.runtime.WorkerSinkTask.poll(WorkerSinkTask.java:322)
        at org.apache.kafka.connect.runtime.WorkerSinkTask.iteration(WorkerSinkTask.java:225)
        at org.apache.kafka.connect.runtime.WorkerSinkTask.execute(WorkerSinkTask.java:193)
        at org.apache.kafka.connect.runtime.WorkerTask.doRun(WorkerTask.java:175)
        at org.apache.kafka.connect.runtime.WorkerTask.run(WorkerTask.java:219)
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
        at java.util.concurrent.FutureTask.run(FutureTask.java:266)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
```
I can tell that the sink is doing something properly as the schema is pulled (see just before the error above) and the table is created successfully in the database with the proper schema:

```
MariaDB [dsb_errors_ksql]> describe DSB_ERROR_TABLE_WINDOWED;
+------------+--------------+------+-----+---------+-------+
| Field      | Type         | Null | Key | Default | Extra |
+------------+--------------+------+-----+---------+-------+
| MOD_CLASS  | varchar(256) | YES  |     | NULL    |       |
| METHOD     | varchar(256) | YES  |     | NULL    |       |
| KSQL_COL_2 | bigint(20)   | YES  |     | NULL    |       |
+------------+--------------+------+-----+---------+-------+
3 rows in set (0.00 sec)

```

Here are the various configs:
Create sink JSON:
```
{ "name": "dev-dsb-errors-mysql-sink",
   "config": { 
		"connector.class": "io.confluent.connect.jdbc.JdbcSinkConnector",
		"tasks.max": "1",
                "topics": "DSB_ERROR_TABLE_WINDOWED", 
		"connection.url": "jdbc:mysql://os-compute-d01.maeagle.corp:32692/dsb_errors_ksql?user=dsb&password=kb2rgb66",
		"auto.create": "true",
		"key.converter": "org.apache.kafka.connect.storage.StringConverter",
		"key.converter.schema.registry.url": "http://kafka-d01.maeagle.corp:8081",
		"value.converter": "io.confluent.connect.avro.AvroConverter",
		"value.converter.schema.registry.url": "http://kafka-d01.maeagle.corp:8081",
		"key.converter.schemas.enable": "false",
		"value.converter.schemas.enable": "false",
		"pk.mode": "none"
   }	   
}
```
ksql table config:

```
ksql> describe extended  DSB_ERROR_TABLE_WINDOWED;

Name                 : DSB_ERROR_TABLE_WINDOWED
Type                 : TABLE
Key field            : KSQL_INTERNAL_COL_0|+|KSQL_INTERNAL_COL_1
Key format           : STRING
Timestamp field      : Not set - using <ROWTIME>
Value format         : AVRO
Kafka topic          : DSB_ERROR_TABLE_WINDOWED (partitions: 4, replication: 1)

 Field      | Type
----------------------------------------
 ROWTIME    | BIGINT           (system)
 ROWKEY     | VARCHAR(STRING)  (system)
 MOD_CLASS  | VARCHAR(STRING)
 METHOD     | VARCHAR(STRING)
 KSQL_COL_2 | BIGINT
----------------------------------------

Queries that write into this TABLE
-----------------------------------
CTAS_DSB_ERROR_TABLE_WINDOWED_11 : create table DSB_ERROR_TABLE_windowed as  select mod_class, method, count(*) from DSB_ERROR_STREAM window session ( 60 seconds)   group by mod_class, method  having count(*) > 0;
```

stream details:
```
ksql> describe extended DSB_ERROR_STREAM;

Name                 : DSB_ERROR_STREAM
Type                 : STREAM
Key field            :
Key format           : STRING
Timestamp field      : Not set - using <ROWTIME>
Value format         : AVRO
Kafka topic          : dev-dsb-errors (partitions: 1, replication: 1)

 Field      | Type
----------------------------------------
 ROWTIME    | BIGINT           (system)
 ROWKEY     | VARCHAR(STRING)  (system)
 TIMESTAMP  | BIGINT
 DSB_SERVER | VARCHAR(STRING)
 ACCOUNT    | VARCHAR(STRING)
 MOD_CLASS  | VARCHAR(STRING)
 METHOD     | VARCHAR(STRING)
 TIMEOUT    | INTEGER
----------------------------------------


```

schema subjects currently in the schema-registry:
```
[
    "_confluent-ksql-default_query_CTAS_DSB_ERROR_TABLE_WINDOWED_11-KSTREAM-AGGREGATE-STATE-STORE-0000000006-repartition-value",
    "divolte-data-value",
    "DSB_ERROR_TABLE_WINDOWED-value",
    "dev-dsb-errors-key",
    "dev-dsb-errors-value",
    "_confluent-ksql-default_query_CTAS_DSB_ERROR_TABLE_WINDOWED_11-KSTREAM-AGGREGATE-STATE-STORE-0000000006-changelog-value"
]
```
My feeling is that there is something not being  set up in the schema-registry as I see a "DSB_ERROR_TABLE_WINDOWED-value" subject  but there is not a "DSB_ERROR_TABLE_WINDOWED-key" subject.  Although I don't know if there should be one...

My impression was that KSQL would handle all the internal details of schema-registry setup when creating an AVRO table from an AVRO stream.

Any help would be appreciated.  I've seen a few issues like this realated to JSON tables, but very little with AVRO tables.

Thanks!

--John

**Comment by fortinj66** [comment]
Additional information:
If I look at the table with the avro-consumer I see this:

```
[root@kafka-d01 kafka]# /bin/kafka-avro-console-consumer --topic "DSB_ERROR_TABLE_WINDOWED" --bootstrap-server "kafka-d01:9092"   --from-beginning --consumer.config /etc/kafka/consumer.properties
null
{"MOD_CLASS":{"string":"TRAINING:NMTSS"},"METHOD":{"string":"getTrainingMeetingSearch"},"KSQL_COL_2":{"long":66}}
null
{"MOD_CLASS":{"string":"TRAINING:NMTSS"},"METHOD":{"string":"getTrainingMeetingSearch"},"KSQL_COL_2":{"long":67}}
null
{"MOD_CLASS":{"string":"TRAINING:NMTSS"},"METHOD":{"string":"getTrainingMeetingSearch"},"KSQL_COL_2":{"long":68}}
```
If I use ksql print I get this:

```
 ksql>print DSB_ERROR_TABLE_WINDOWED from beginning
12/8/18 1:57:44 PM EST, TRAINING:NMTSS|+|getTrainingMeetingSearchg�,e+, {"MOD_CLASS": "TRAINING:NMTSS", "METHOD": "getTrainingMeetingSearch", "KSQL_COL_2": 52}
12/8/18 1:57:47 PM EST, TRAINING:NMTSS|+|getTrainingMeetingSearchg�,e+, {"MOD_CLASS": "TRAINING:NMTSS", "METHOD": "getTrainingMeetingSearch", "KSQL_COL_2": 53}
12/8/18 1:57:50 PM EST, TRAINING:NMTSS|+|getTrainingMeetingSearchg�,e+, {"MOD_CLASS": "TRAINING:NMTSS", "METHOD": "getTrainingMeetingSearch", "KSQL_COL_2": 54}
Topic printing ceased
```
It looks like the null value should have the rowid and rowkey in it and it doesn't...

**Comment by rmoff** [comment]
As an aside: 

> there is not a "DSB_ERROR_TABLE_WINDOWED-key" subject

KSQL only supports STRING keys, so you wouldn't see an Avro schema for the key registered.

**Comment by rmoff** [comment]
This is a problem with the JDBC sink, not KSQL per se. The key written, whilst a bunch of binary junk so far as a string key is concerned, is valid when used in other sinks (e.g. Elasticsearch). It seems to be the JDBC Sink is choking on something in the weeds. I've raised https://github.com/confluentinc/kafka-connect-jdbc/issues/591. 

For others hitting this issue, setting `"pk.mode": "kafka"` in the sink config will work.

**Comment by k1th** [comment]
Using `"pk.mode": "kafka"` won't work for updates, because UPDATES to windows will be per `__connect_topic,__connect_partition,__connect_offset` and you're going to have multiple rows in the DB for the same record_key.

I used     
`"key.converter": "org.apache.kafka.connect.converters.ByteArrayConverter",`
which in my tests works for Postgres. Should work for all databases that support primary keys on that kind of column.
